# Your grades and GPA at Illinois!

In this lab, you will explore the GPA dataset and find out about the easiest, hardest, and most popular courses at Illinois!

## Part 0 | Loading the Dataset

For the first part of our lab, we'll be using the GPA dataset we saw in class.

You can download the dataset by following to following link to the github storing the dataset: https://raw.githubusercontent.com/wadefagen/datasets/master/gpa/uiuc-gpa-dataset.csv

To download this file, you will want to **right click** and click **Save As**.

In [1]:
import pandas as pd
df = pd.read_csv("gpa.csv")

## Part 1 | Finding Total GPAs

Add two new columns, `Total Students` and `Average GPA`, that compute the total number of students in each course and the average weighted GPA for each course.

You should not include students with a `W` in either column and feel free to add other columns to help you out!

In [4]:
df["Total Students"] = df["A+"] + df["A"] + df["A-"] + df["B+"] + df["B"] + df["B-"] + df["C+"] + df["C"] + df["C-"] + df["D+"] + df["D"] + df["D-"] + df["F"]
df["Average GPA"] = (df["A+"]*4 + df["A"]*4 + df["A-"]*3.67 + df["B+"]*3.33 + df["B"]*3 + df["B-"]*2.67 + df["C+"]*2.33 + df["C"]*2 + df["C-"]*1.67 + df["D+"]*1.33 + df["D"]*1 + df["D-"]*0.67) / df["Total Students"]

In [5]:
## == TEST CASES for Puzzle 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.

assert(df[
  (df["YearTerm"] == "2010-su") & \
  (df["Subject"] == "STAT") & \
  (df["Number"] == 410) \
]["Total Students"].sum() == 31), "It appears your \"Total Students\" column is incorrect."

assert(df[
  (df["YearTerm"] == "2018-fa") & \
  (df["Subject"] == "CS") & \
  (df["Number"] == 225) \
]["Total Students"].sum() == 697), "It appears your \"Total Students\" column is incorrect."

assert( abs(df[
  (df["YearTerm"] == "2018-fa") & \
  (df["Subject"] == "STAT") & \
  (df["Number"] == 100) \
]["Average GPA"].mean() - 3.247) < 0.01), "It appears your \"Average GPA\" column is incorrect."

## Part 2 | Finding the easiest and hardest courses

### Puzzle 2.1: Hardest Courses in the Dataset

Using functions you already know, find the **30 hardest courses** at Illinois based on their "Average GPA".

In [6]:
df_hardestCourses = df.nsmallest(30, "Average GPA")
df_hardestCourses



Year    Term YearTerm Subject  Number                   Course Title  \
1725   2018    Fall  2018-fa     LER     120    Contemporary Labor Problems   
14173  2016  Spring  2016-sp     LER     100  Introduction to Labor Studies   
21986  2014    Fall  2014-fa     LER     120    Contemporary Labor Problems   
24254  2014  Spring  2014-sp     LER     120    Contemporary Labor Problems   
27051  2013    Fall  2013-fa    MATH       2           Introductory Algebra   
24329  2014  Spring  2014-sp    MATH     220                       Calculus   
11901  2016    Fall  2016-fa     LER     120    Contemporary Labor Problems   
12169  2016    Fall  2016-fa     MCB     245     Human Anat & Physiol Lab I   
22033  2014    Fall  2014-fa     LLS     296  Bilingual US Latina/o Culture   
22197  2014    Fall  2014-fa     MCB     245     Human Anat & Physiol Lab I   
9102   2017  Spring  2017-sp     LER     100  Introduction to Labor Studies   
3076   2018  Spring  2018-sp    CHEM     101         Introductory Chemistry   
11897  2016    Fall  2016-fa     LER     100  Introduction to Labor Studies   
11900  2016    Fall  2016-fa     LER     110     Labor and Social Movements   
37521  2011  Spring  2011-sp    MATH     115       Preparation for Calculus   
12160  2016    Fall  2016-fa     MCB     245     Human Anat & Physiol Lab I   
37457  2011  Spring  2011-sp     LER     100  Introduction to Labor Studies   
23849  2014  Spring  2014-sp    ENGL     274          Contemporary Classics   
16964  2015    Fall  2015-fa    MATH     181           A Mathematical World   
1935   2018    Fall  2018-fa     MCB     245     Human Anat & Physiol Lab I   
12165  2016    Fall  2016-fa     MCB     245     Human Anat & Physiol Lab I   
35869  2011    Fall  2011-fa     TAM     251   Introductory Solid Mechanics   
41077  2011  Summer  2011-su     TAM     251   Introductory Solid Mechanics   
21782  2014    Fall  2014-fa      IE     300               Analysis of Data   
29359  2013  Spring  2013-sp    MATH     115       Preparation for Calculus   
34982  2011    Fall  2011-fa     LER     110     Labor and Social Movements   
40190  2011  Summer  2011-su     LER     110     Labor and Social Movements   
9107   2017  Spring  2017-sp     LER     110     Labor and Social Movements   
14177  2016  Spring  2016-sp     LER     110     Labor and Social Movements   
12166  2016    Fall  2016-fa     MCB     245     Human Anat & Physiol Lab I   

       A+  A  A-  B+     ...        C  C-  D+   D  D-   F  W  \
1725    0  0   1   2     ...        1   2   1   1   0  15  0   
14173   0  1   0   1     ...        4   1   1   1   1  10  0   
21986   0  4   0   1     ...        4   1   0   1   0  15  1   
24254   0  1   3   0     ...        3   0   0   0   2  11  0   
27051   0  0   0   0     ...       10   5   3   3   2   5  3   
24329   0  3   0   0     ...        5   0   2   1   0  11  1   
11901   0  2   2   1     ...        0   1   2   1   2   8  2   
12169   0  0   0   0     ...        3   4   1   2   0   6  1   
22033   3  1   0   1     ...        1   1   1   0   2  11  1   
22197   0  0   1   0     ...        1   3   5   1   1   5  0   
9102    0  2   1   1     ...        3   2   0   1   0   9  0   
3076    1  5   8   1     ...       10   9  11  23  10  26  7   
11897   0  2   1   2     ...        3   2   0   1   1  10  0   
11900   0  4   2   0     ...        2   1   0   1   4   8  0   
37521   1  4   2   1     ...        5   0   9  11   2  18  0   
12160   0  0   1   1     ...        1   3   3   4   1   4  2   
37457   2  1   0   2     ...        1   2   0   0   1   9  2   
23849   0  1   3   0     ...        1   2   1   0   0   8  1   
16964   1  3   0   2     ...        3   1   1   1   1   8  0   
1935    0  0   1   2     ...        4   1   0   1   1   7  0   
12165   0  0   1   2     ...        3   1   1   1   3   6  0   
35869   1  0   4   2     ...       23   3   6  23   9  11  0   
41077   1  0   4   2     ...       23   3   6  23   9  11  0   
21782   4  1   3   1 

### Puzzle 2.2: Hardest Courses in 2018

Find the **30 hardest courses in 2018** at Illinois based on their "Average GPA".

In [8]:
df_2018 = df[df["Year"]==2018]
df_hardestCourses_2018 = df_2018.nsmallest(30, "Average GPA")
df_hardestCourses_2018



Year    Term YearTerm Subject  Number                    Course Title  \
1725  2018    Fall  2018-fa     LER     120     Contemporary Labor Problems   
3076  2018  Spring  2018-sp    CHEM     101          Introductory Chemistry   
1935  2018    Fall  2018-fa     MCB     245      Human Anat & Physiol Lab I   
4168  2018  Spring  2018-sp    MATH     446       Applied Complex Variables   
4871  2018  Summer  2018-su    CHEM     102             General Chemistry I   
4095  2018  Spring  2018-sp    MATH     112                         Algebra   
1723  2018    Fall  2018-fa     LER     100   Introduction to Labor Studies   
633   2018    Fall  2018-fa    CHEM     104            General Chemistry II   
379   2018    Fall  2018-fa    ASTR     404            Stellar Astrophysics   
1238  2018    Fall  2018-fa     FIN     300               Financial Markets   
4232  2018  Spring  2018-sp     MCB     250              Molecular Genetics   
1928  2018    Fall  2018-fa     MCB     245      Human Anat & Physiol Lab I   
2173  2018    Fall  2018-fa    PSYC     100                     Intro Psych   
1724  2018    Fall  2018-fa     LER     110      Labor and Social Movements   
1073  2018    Fall  2018-fa    ECON     302      Inter Microeconomic Theory   
1936  2018    Fall  2018-fa     MCB     245      Human Anat & Physiol Lab I   
4870  2018  Summer  2018-su    CHEM     102             General Chemistry I   
3565  2018  Spring  2018-sp    ECON     302      Inter Microeconomic Theory   
1501  2018    Fall  2018-fa      IE     310   Determin Models in Optmzation   
649   2018    Fall  2018-fa    CHEM     232  Elementary Organic Chemistry I   
2761  2018  Spring  2018-sp      AE     321   Mechs of Aerospace Structures   
1720  2018    Fall  2018-fa     LER     100   Introduction to Labor Studies   
3564  2018  Spring  2018-sp    ECON     302      Inter Microeconomic Theory   
2301  2018    Fall  2018-fa     RST     410       Strategic Thinking in RST   
1814  2018    Fall  2018-fa    MATH     115        Preparation for Calculus   
3277  2018  Spring  2018-sp     CMN     310        The Rhetorical Tradition   
1587  2018    Fall  2018-fa     KIN     140    Social Sci of Human Movement   
3599  2018  Spring  2018-sp    ECON     480    Industrial Comp and Monopoly   
4096  2018  Spring  2018-sp    MATH     115        Preparation for Calculus   
3571  2018  Spring  2018-sp    ECON     302      Inter Microeconomic Theory   

      A+   A  A-  B+     ...        C  C-  D+   D  D-   F  W  \
1725   0   0   1   2     ...        1   2   1   1   0  15  0   
3076   1   5   8   1     ...       10   9  11  23  10  26  7   
1935   0   0   1   2     ...        4   1   0   1   1   7  0   
4168   2   2   1   5     ...        2   0   3   4   6   9  1   
4871   3   1   1   4     ...        6   3   3   3   3  13  2   
4095   0   1   1   1     ...        3   3   2   1   0   4  1   
1723   0   3   3   1     ...        1   3   1   0   0   9  2   
633   10   9   9  14     ...       16  23  25  17  11  20  4   
379    0   3   0   3     ...        4   5   3   4   0   4  0   
1238   0   1   1   1     ...        5   2   1   0   2   3  1   
4232   0   1   1   3     ...        2   3   1   1   1   3  0   
1928   1   0   1   3     ...        2   2   3   2   2   1  1   
2173   0   4   1   0     ...       10   3   1   9   1   2  0   
1724   0   3   4   0     ...        6   0   1   3   1   5  0   
1073   2   4   1   0     ...        1   5   3   2   2   4  3   
1936   0   0   2   3     ...        1   1   1   1   2   3  0   
4870   0   2   4   3     ...        5   6   4   0   5   2  0   
3565   1   3   3   3     ...        5   2   1   2   0   6  0   
1501   1   9   6   4     ...        1   1   9   6   9   6  0   
649    1   2   1   0     ...        6   0   0   6   0   1  0   
2761   1   8   4   5     ...        3   2   1   2   3   9  0   
1720   0   8   3   0     ...        0   0   0   0   1   9  0   
3564   3   4   0   1     ...        5   3   4   2   0   3  1   
2301   0   0   2   5

### Puzzle 2.3: Easiest Courses in 2018

Find the **30 easiest courses in 2018** at Illinois based on their "Average GPA".

In [9]:
df_easiestCourses_2018 = df_2018.nlargest(30, "Average GPA")
df_easiestCourses_2018



Year    Term YearTerm Subject  Number                    Course Title  \
3688  2018  Spring  2018-sp    EPSY     490  Prep for Recruitment Counselor   
1124  2018    Fall  2018-fa    EDUC     101   Education Orientation Seminar   
129   2018    Fall  2018-fa    ACCY     518       Financial Statement Fraud   
2044  2018    Fall  2018-fa     MUS     269                  Wind Orchestra   
2934  2018  Spring  2018-sp    BADM     420   Advanced Marketing Management   
1486  2018    Fall  2018-fa      IB     199      Undergraduate Open Seminar   
2436  2018    Fall  2018-fa    SPED     405     Gen Educator's Role in SPED   
4914  2018  Summer  2018-su    EPSY     408    Learning & Hum Dev w/ EdTech   
2792  2018  Spring  2018-sp    ANSC     103       Working With Farm Animals   
2998  2018  Spring  2018-sp    BIOE     436                Senior Design II   
1662  2018    Fall  2018-fa     LAS     101                Freshman Seminar   
860   2018    Fall  2018-fa      CS     498               Digital Forensics   
707   2018    Fall  2018-fa      CI     406    Thry Prac in Elem Schl Tch I   
3422  2018  Spring  2018-sp     ECE     206  Elec & Electronic Circuits Lab   
3771  2018  Spring  2018-sp    FSHN     419      Food Ingredient Technology   
1140  2018    Fall  2018-fa     ENG     198                 IEFX Projects 1   
3202  2018  Spring  2018-sp    CHLH     380       Orientation to Internship   
3390  2018  Spring  2018-sp      CS     546         Machine Learning in NLP   
4992  2018  Summer  2018-su    SOCW     531   Field Practicum and Seminar I   
4665  2018  Spring  2018-sp    SOCW     402                     Practice II   
2340  2018    Fall  2018-fa     SHS     270    Comm Disability in the Media   
3940  2018  Spring  2018-sp      IS     404  Lit and Resources Young Adults   
2894  2018  Spring  2018-sp    BADM     320         Principles of Marketing   
4511  2018  Spring  2018-sp    PSYC     396              The Attentive Mind   
4913  2018  Summer  2018-su    EPSY     407  Adult Learning and Development   
4759  2018  Spring  2018-sp      TE     460  Lect in Engrg Entrepreneurship   
393   2018    Fall  2018-fa    BADM     261       Technology & Mgmt Seminar   
1498  2018    Fall  2018-fa      IB     534          Evolution and Medicine   
2955  2018  Spring  2018-sp    BADM     520            Marketing Management   
3643  2018  Spring  2018-sp     ENG     398         ELA Leadership Training   

      A+    A  A-  B+     ...       C  C-  D+  D  D-  F  W  \
3688  43   37   0   0     ...       0   0   0  0   0  0  0   
1124   0   67   0   0     ...       0   0   0  0   0  0  1   
129   32    8   0   1     ...       0   0   0  0   0  0  0   
2044   0   54   1   1     ...       0   0   0  0   0  0  1   
2934   0   54   0   0     ...       0   0   0  0   0  0  0   
1486   0  146   0   0     ...       0   0   0  0   0  0  0   
2436   0   48   0   0     ...       0   0   0  0   0  0  0   
4914  65    2   2   0     ...       0   0   0  0   0  0  1   
2792   0   66   0   1     ...       0   0   0  0   0  0  0   
2998  24   38   3   1     ...       0   0   0  0   0  0  0   
1662  21    4   0   1     ...       0   0   0  0   0  0  0   
860    0   37   0   0     ...       0   0   0  0   0  0  0   
707   14   10   0   1     ...       0   0   0  0   0  0  0   
3422   3   21   0   1     ...       0   0   0  0   0  0  0   
3771   0   24   0   1     ...       0   0   0  0   0  0  0   
1140   0   47   1   0     ...       0   0   0  0   0  0  0   
3202   0  103   4   1     ...       0   0   0  0   0  0  0   
3390   0   80   2   1     ...       0   0   0  0   0  0  1   
4992   0   23   0   1     ...       0   0   0  0   0  0  0   
4665  18    4   0   1     ...       0   0   0  0   0  0  0   
2340   0   98   0   0     ...       1   0   0  0   0  0  0   
3940   0   21   0   1     ...       0   0   0  0   0  0  0   
2894   0  131   6   2     ...       0   0   0  0   0  0  0   
4511  28    2   1   1     ...       0   0   0  0   0  0  0   
4913  

In [10]:
## == TEST CASES for Puzzle 2 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error or output, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, RE-RUN your code, and then RE-RUN this cell.

assert( len(df_hardestCourses) == 30 ), "You should only have the 30 hardest courses."
assert( abs(df_hardestCourses["Average GPA"].mean() - 1.58) < 0.01 ), "Are you sure you have the 30 hardest courses?"

assert( len(df_hardestCourses_2018) == 30 ), "You should only have the 30 hardest courses in 2018."
assert( abs(df_hardestCourses_2018["Average GPA"].mean() - 2.07) < 0.01 ), "Are you sure you have the 30 hardest courses of 2018?"

assert( len(df_easiestCourses_2018) == 30 ), "You should only have the 30 easiest courses in 2018."
assert( abs(df_easiestCourses_2018["Average GPA"].mean() - 3.97) < 0.01 ), "Are you sure you have the 30 easiest courses of 2018?"

## Puzzle 3: Finding Data on  Every Subject

### Puzzle 3.1: Total Grades Given in Each Subject

Using `groupby` and `agg`, find the total number of each grade (eg: "A+"s, "A"s, "A-"s, etc.) for every subject in the DataFrame `df_grouped_subject`.

In [22]:



df_grouped = df.groupby("Subject")
df_grouped_subject = df_grouped.agg("sum").reset_index()
df_grouped_subject

Subject     Year  Number     A+      A     A-     B+      B    B-    C+  \
0       AAS   298016   19199    571   1726    714    460    422   198   116   
1       ABE   229611   38135    254   2200    721    632    827   210   116   
2      ACCY  3798626  689840   5437  18283  12947  12596  11368  4783  2267   
3       ACE  1492356  200528   2207  10864   4329   3698   5811  2264  1403   
4      ACES   245691   14066    569   1708    215     85    235    56    58   
5       ADV   698959  132912   1991   8075   3814   2826   4608  1260   806   
6        AE   570044   96822   1470   4726   2581   1803   2309   991   632   
7      AFRO   193325   19081    305    995    494    288    341   186   106   
8      AFST    20136    2172     58    160     93     42     28    13     4   
9      AGCM    52378    6805     40    309    103     67     74    19     5   
10     AGED   249725   40727    931   2332    957    569    467   212   132   
11      AHS    54360    7096    142    708     37     21     52     4     0   
12      AIS   116761    7049    168    653    261    174    258    67    30   
13     ANSC   966777  139858   3345  14417   1516   1913   5120   825   765   
14     ANTH  1057057   94121   5095  12983   5483   3477   5204  2124  1175   
15     ARAB    50334    5531    287    269     92     68     66    29    18   
16     ARCH   678558  124154   1943   5247   3963   3487   3795  1645   918   
17      ART   193395   16288    287   1503    640    323    217   105    44   
18     ARTD   326269   49617    386   1428   1061    896    703   315   132   
19     ARTE     8063    1079      9     58     10      5      4     3     0   
20     ARTF    42275    2133    114    552    214    130    123    47    37   
21     ARTH   239605   28487    163    729    775    590    687   401   169   
22     ARTS    38227    6695     23    176    101     65     44    28     8   
23     ASRM    18162    3073     42    116     77     55     53    19    25   
24     ASTR   324309   32016   1684   5466   2594   2034   3117  1195   930   
25     ATMS   358442   39968   8899   6905   2493   2147   2984   939   601   
26      AVI    32177    5017     30     84     72     58     66    23    16   
27     BADM  3478560  755852  12670  38741  20979  13229  10529  4090  2070   
28      BCS     8062     460     73     71      9      4     12     5     2   
29     BIOC    56407   12672     72    234    154    155    152    78    51   
..      ...      ...     ...    ...    ...    ...    ...    ...   ...   ...   
131    RHET   780287   48352    360   2979   1986   1222    925   343   151   
132    RLST   917950   58878    916   3936   2172   1735   2302   989   538   
133    RSOC     6034     330      0     83      0      0     27     0     0   
134     RST  1019142  135068   3438  12329   3999   2922   6114  1784  1084   
135    RUSS    16127    1542      9     68     26     23     31     9     7   
136    SAME     6039     437      2     29     30     20     31    12    14   
137     SBC    18161    4542    124     99     41     19      9     6     2   
138    SCAN   100697   12780     83    472    272    194    243   114    65   
139     SCR     4027     230     28     40     11      5      8     3     3   
140      SE    64566   11159    223    476    275    216    261    91    52   
141     SHS  1212440  165877   2355  16249   3193   2160   7016  1397   832   
142    SLAV    16097     945      2    111     27     46     46     4     0   
143    SLCL     6053     600     21     46     18      7     10     0     2   
144     SOC  1176069  114613   1677   7128   3764   2273   2578  1318   723   
145    SOCW  1124044  241488   1750   7133   2751   1694   1730   599   282   
146    SPAN   615664   63435    862   1720   1299    938    912   475   365   
147    SPED   392602   67425   1062   9342    496    324   1560   108    39   
148    STAT  1210775  204459   7854  19509   8856   6531   6963  3603  2207   
149    SWAH    10058    1412     24     21     15     

In [23]:
## == TEST CASES for Puzzle 3.1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, RE-RUN your code, and then RE-RUN this cell.

assert( len(df_grouped_subject) == 161 ), "Are you sure you grouped by subjects?"
assert( df_grouped_subject[ df_grouped_subject["Subject"] == "CS" ]["A"].sum() == 35107 ), "Are you sure you counted the right number of each subject?"

### Puzzle 3.2: Overall Average GPA

In `df_grouped_subject`, add a new column called `"Overall Average GPA"` that computes the total overall GPA of each subject at Illinois.

In [24]:
df_grouped_subject["Overall Average GPA"] = (df_grouped_subject["A+"]*4 + df_grouped_subject["A"]*4 + df_grouped_subject["A-"]*3.67 + df_grouped_subject["B+"]*3.33 + df_grouped_subject["B"]*3 + df_grouped_subject["B-"]*2.67 + df_grouped_subject["C+"]*2.33 + df_grouped_subject["C"]*2 + df_grouped_subject["C-"]*1.67 + df_grouped_subject["D+"]*1.33 + df_grouped_subject["D"]*1 + df_grouped_subject["D-"]*0.67) / df_grouped_subject["Total Students"]
df_grouped_subject



Subject     Year  Number     A+      A     A-     B+      B    B-    C+  \
0       AAS   298016   19199    571   1726    714    460    422   198   116   
1       ABE   229611   38135    254   2200    721    632    827   210   116   
2      ACCY  3798626  689840   5437  18283  12947  12596  11368  4783  2267   
3       ACE  1492356  200528   2207  10864   4329   3698   5811  2264  1403   
4      ACES   245691   14066    569   1708    215     85    235    56    58   
5       ADV   698959  132912   1991   8075   3814   2826   4608  1260   806   
6        AE   570044   96822   1470   4726   2581   1803   2309   991   632   
7      AFRO   193325   19081    305    995    494    288    341   186   106   
8      AFST    20136    2172     58    160     93     42     28    13     4   
9      AGCM    52378    6805     40    309    103     67     74    19     5   
10     AGED   249725   40727    931   2332    957    569    467   212   132   
11      AHS    54360    7096    142    708     37     21     52     4     0   
12      AIS   116761    7049    168    653    261    174    258    67    30   
13     ANSC   966777  139858   3345  14417   1516   1913   5120   825   765   
14     ANTH  1057057   94121   5095  12983   5483   3477   5204  2124  1175   
15     ARAB    50334    5531    287    269     92     68     66    29    18   
16     ARCH   678558  124154   1943   5247   3963   3487   3795  1645   918   
17      ART   193395   16288    287   1503    640    323    217   105    44   
18     ARTD   326269   49617    386   1428   1061    896    703   315   132   
19     ARTE     8063    1079      9     58     10      5      4     3     0   
20     ARTF    42275    2133    114    552    214    130    123    47    37   
21     ARTH   239605   28487    163    729    775    590    687   401   169   
22     ARTS    38227    6695     23    176    101     65     44    28     8   
23     ASRM    18162    3073     42    116     77     55     53    19    25   
24     ASTR   324309   32016   1684   5466   2594   2034   3117  1195   930   
25     ATMS   358442   39968   8899   6905   2493   2147   2984   939   601   
26      AVI    32177    5017     30     84     72     58     66    23    16   
27     BADM  3478560  755852  12670  38741  20979  13229  10529  4090  2070   
28      BCS     8062     460     73     71      9      4     12     5     2   
29     BIOC    56407   12672     72    234    154    155    152    78    51   
..      ...      ...     ...    ...    ...    ...    ...    ...   ...   ...   
131    RHET   780287   48352    360   2979   1986   1222    925   343   151   
132    RLST   917950   58878    916   3936   2172   1735   2302   989   538   
133    RSOC     6034     330      0     83      0      0     27     0     0   
134     RST  1019142  135068   3438  12329   3999   2922   6114  1784  1084   
135    RUSS    16127    1542      9     68     26     23     31     9     7   
136    SAME     6039     437      2     29     30     20     31    12    14   
137     SBC    18161    4542    124     99     41     19      9     6     2   
138    SCAN   100697   12780     83    472    272    194    243   114    65   
139     SCR     4027     230     28     40     11      5      8     3     3   
140      SE    64566   11159    223    476    275    216    261    91    52   
141     SHS  1212440  165877   2355  16249   3193   2160   7016  1397   832   
142    SLAV    16097     945      2    111     27     46     46     4     0   
143    SLCL     6053     600     21     46     18      7     10     0     2   
144     SOC  1176069  114613   1677   7128   3764   2273   2578  1318   723   
145    SOCW  1124044  241488   1750   7133   2751   1694   1730   599   282   
146    SPAN   615664   63435    862   1720   1299    938    912   475   365   
147    SPED   392602   67425   1062   9342    496    324   1560   108    39   
148    STAT  1210775  204459   7854  19509   8856   6531   6963  3603  2207   
149    SWAH    10058    1412     24     21     15     

In [25]:
## == TEST CASES for Puzzle 3.2 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error or output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, RE-RUN your code, and then RE-RUN this cell.

assert( abs(df_grouped_subject[ df_grouped_subject["Subject"] == "CS" ]["Overall Average GPA"].sum() - 3.236) < 0.01 ), "Your overall average GPA does not appear correct."

### Puzzle 3.3: Find the 15 easiest subjects at Illinois

Finally, find the 15 easiest subjects at Illinois based on their "Overall Average GPA"

In [28]:
df_easiest_subjects = df_grouped_subject.nlargest(30, "Overall Average GPA")
df_easiest_subjects



Subject    Year  Number    A+     A    A-   B+     B   B-  C+    C  C-  \
34       CB    4022    1250     0   140     0    0    17    0   0    1   0   
40      CHP   12102    1194    42    94    15    8     5    0   0    3   0   
155     UKR    8048     452     5   190    43   20     8    4   0    1   0   
128    REES    6047     603    32    34     7    2     5    2   0    0   0   
53     EDUC  106864    8974   307  1688   220  102    92   23  16   28   4   
33      BUS  493460   26919   493  4525   924  413   230   82  31   34   9   
19     ARTE    8063    1079     9    58    10    5     4    3   0    1   0   
137     SBC   18161    4542   124    99    41   19     9    6   2    2   0   
59      EOL   82576   21905    62   866   188  101    69    8   2    6   0   
126     PSM   16126    4013   121    37    13    4    10    4   2    3   2   
11      AHS   54360    7096   142   708    37   21    52    4   0   14   2   
89       IS  829740  214129  1211  6353  1969  774   535  159  40   49  13   
41       CI  936346  202539  1381  7521  2000  912   630  173  76   81  27   
60     EPOL    6053    1650     1    94     9    2     2    1   0    1   1   
151      TE   80670   15049   302   663   170   85    61    9   9   18   6   
95      KOR   30208    3043   130   104    42   23    21    6   4    1   1   
93       JS    2018     355     4     7     7    2     1    0   0    0   0   
83      HRD   38306    8838    61   291    72   43    27   11   3    2   0   
54      EIL   68497   15106   123   436   139   65    55   10  10    6   5   
147    SPED  392602   67425  1062  9342   496  324  1560  108  39  338  14   
61      EPS  310055   51965   600  2644   990  355   213   95  38   31  20   
123     POL   16109     920    25   114   112   37    17    2   0    0   0   
28      BCS    8062     460    73    71     9    4    12    5   2    6   4   
31     BIOP    6032    1203    13    45     6    5     3    2   0    3   0   
55      ENG  483264   75244  1807  4731  1217  584   662  169  94  141  37   
139     SCR    4027     230    28    40    11    5     8    3   3    1   0   
87     IHLT  102807   14031   514  1285   133   51   162   21  14   58  10   
49     DANC  408824   32576  1342  2817   866  462   328  171  79   85  37   
65     EURO   14101    2645    11   108    46   12     4    4   2    0   0   
143    SLCL    6053     600    21    46    18    7    10    0   2    0   1   

     D+   D  D-    F   W  Total Students  Average GPA  Overall Average GPA  
34    0   0   0    0   2             158     7.767925             3.879747  
40    0   0   0    1   0             168    23.124832             3.849345  
155   0   0   0    0   0             271    15.181337             3.841661  
128   0   1   0    0   0              83    11.434037             3.827590  
53    2  11   2    8   4            2503   199.988373             3.827179  
33    4   9   2   23  10            6779   936.301713             3.823312  
19    0   0   0    0   0              90    15.232428             3.815111  
137   1   1   0    0   0             304    34.289368             3.814967  
59    0   4   0    6   3            1312   155.348510             3.801303  
126   0   0   0    1   0             197    30.340307             3.795482  
11    0  10   0   15   0            1005   102.790092             3.794766  
89    2  17   1   31  14           11154  1564.480532             3.794372  
41    7  11   0   35  19           12854  1760.424735             3.791921  
60    0   0   0    3   0             114    11.462912             3.789737  
151   0   1   0    8   5            1332   150.993088             3.785263  
95    0   0   0    1   1             333    56.670767             3.780000  
93    0   0   0    0   1              21     3.778571             3.778571  
83    0   2   0    2   1             514    71.574511             3.771965  
54    0   2   0    6   5             857   126.682236             3.733874  
147   5  85   0  122  21           13495   73

### Puzzle 3.4: Find the hardest 15 subjects at Illinois

Find the hardest 15 subjects at Illinois based on their "Overall Average GPA"

In [29]:
df_hardest_subjects = df_grouped_subject.nsmallest(30, "Overall Average GPA")
df_hardest_subjects



Subject     Year  Number     A+      A     A-     B+      B     B-     C+  \
159      VM   161151   48575      0   1503      8      0   5029      1      0   
150     TAM   521580   83139   1611   3910   3026   3022   3722   2472   1982   
105    MATH  3951142  606935   9720  25074  16157  16442  22554  13336  11478   
121    PHYS  1409784  202288  11513  13751   9524   9157   9720   8607   7085   
99      LAT    12076     613     14     20     17     16     21      7     14   
136    SAME     6039     437      2     29     30     20     31     12     14   
36     CHBE   425030   88102    511   2952   1487   1565   2895   1410   1136   
38     CHIN     4023     445      0      8      9      4      9      3      1   
37     CHEM  3126162  282872   9759  38674  15498  15847  21077  12242  11048   
107     MCB  3256457  414467   5797  11913   9339   9153  11042   6904   5166   
125      PS  1081252  145728   1748   6582   5011   4236   5534   2872   1835   
51      ECE  2934499  505822   6636  18872  10654   9629  12525   6792   4765   
73     GEOL   489178   42561   1448   2179   2135   1992   1758   1386    834   
52     ECON  5535028  663026  11201  23537  13340  12811  17276   7704   5686   
127    PSYC  2765150  316661   6306  22977   8349   8556  17413   5428   3981   
154     TSM   304136   47830    314   1628    674    610   1353    402    206   
120    PHIL   930169   73514    533   3100   2855   2438   2316   1443    759   
101     LER  1322760  201816    928   6590   3218   1780   2278    930    443   
24     ASTR   324309   32016   1684   5466   2594   2034   3117   1195    930   
44      CMN  2345248  308310   2329  13687   8336   6614   8448   3909   2618   
135    RUSS    16127    1542      9     68     26     23     31      9      7   
21     ARTH   239605   28487    163    729    775    590    687    401    169   
90     ITAL   114659    8145      6    553     74     76    386     15      3   
116    NRES   434989   70043   1759   4515   3202   2566   3123   1529    916   
86       IE   412969   83689    816   2855   1984   1699   1906    923    552   
91     JAPN    28182    2826     35     70     63     33     23     28     12   
112     MSE   660589  127734   1280   4462   2972   2637   2680   1494    924   
149    SWAH    10058    1412     24     21     15     17      8     10      5   
146    SPAN   615664   63435    862   1720   1299    938    912    475    365   
78      GWS   312114   32715    164   1230    844    538    605    293    149   

         C    C-    D+     D    D-     F     W  Total Students  Average GPA  \
159   2916     0     0   274     0     8     8            9739   226.664769   
150   2056  1315   797   996   460  1093   148           26462   766.988665   
105  13799  7051  4048  5107  2105  6563  1323          153434  5794.147913   
121   6959  5199  3367  2525  2887  1627   655           91921  2137.086474   
99       6     3     1     3     2     8     0             132    17.649451   
136     17     5     2     2     3     5     1             172     8.849293   
36    1854   631   146   301    54   240    64           15182   648.109073   
38       3     1     1     1     1     1     0              42     6.017143   
37   13343  7108  3388  3713  1877  3384   987          156958  5078.958910   
107   4879  2840  1596  1274   747  1116   327           71766  4996.612722   
125   2000   895   381   531   202  1025   180           32852  1691.091207   
51    6137  3112  1076  2057   673  1976   400           84904  4630.059529   
73     677   426   267   196   133   364    56           13795   759.148730   
52    7517  3089  1970  2591  1106  2906   640          110734  8413.231133   
127   8046  1781  1063  2080   455  1933   448           88368  4327.363977   
154    522   134    31   140    56   137    42            6207   473.201745   
120    616   326   144   176   138   481    91           15325  1450.222816   
101    857   391   181   333   189  1385   132           19503  2020.4

### Puzzle 3.5: Find your major!

Finally, find your major/subject in the DataSet and check out how hard it is!

In [33]:
df_ECE = df_grouped_subject[df_grouped_subject["Subject"] == "ECE"]
df_ECE["Overall Average GPA"]

51    3.078994
Name: Overall Average GPA, dtype: float64

- The overall average GPA for ECE is 3.078994

# Finale!

You're almost done -- congratulations!

You need to do two more things:

1. Save your work.  To do this, create a **notebook checkpoint** by using the menu within the notebook to go **File -> Save and Checkpoint**

2. After you have saved and checkpointed, exit this notebook, and return to the Data Science DISCOVERY page on how to use git to turn this notebook into the course!